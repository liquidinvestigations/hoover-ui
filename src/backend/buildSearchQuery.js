import { DateTime } from 'luxon'
import { DEFAULT_FACET_SIZE, DEFAULT_INTERVAL, DEFAULT_OPERATOR, HIGHLIGHT_SETTINGS, PRIVATE_FIELDS } from '../constants/general'
import { daysInMonth } from '../utils'
import { aggregationFields } from '../constants/aggregationFields'

const expandPrivate = (field, uuid) => {
    if (PRIVATE_FIELDS.includes(field)) {
        return `${field}.${uuid}`
    }
    return field
}

const buildQuery = (q, filters, searchFields) => {
    const qs = {
        query_string: {
            query: q,
            default_operator: DEFAULT_OPERATOR,
            fields: searchFields.all,
            lenient: true,
        },
    }

    const ranges = []
    ;['date', 'date-created'].forEach((field) => {
        const value = filters[field]
        if (value?.from && value?.to) {
            ranges.push({
                range: {
                    [field]: {
                        gte: value.from,
                        lte: value.to,
                    },
                },
            })
        }
    })

    if (ranges.length) {
        return {
            bool: {
                must: [qs, ...ranges],
            },
        }
    }

    return qs
}

const buildSortQuery = (order) =>
    order
        ?.reverse()
        .map(([field, direction = 'asc']) =>
            field.startsWith('_') ? { [field]: { order: direction } } : { [field]: { order: direction, missing: '_last' } }
        ) || []

const buildTermsField = (field, uuid, terms, page = 1, size = DEFAULT_FACET_SIZE) => {
    const fieldKey = expandPrivate(field, uuid)

    let filterClause = null
    if (terms?.include?.length) {
        if (aggregationFields[field].type === 'term-and') {
            filterClause = terms.include.map((term) => ({
                term: { [fieldKey]: term },
            }))
        } else {
            filterClause = {
                terms: { [fieldKey]: terms.include },
            }
        }
    }

    let filterExclude = null
    if (terms?.exclude?.length) {
        filterExclude = {
            terms: { [fieldKey]: terms.exclude },
        }
    }

    let filterMissing = null
    if (terms?.missing === 'true') {
        filterMissing = false
    } else if (terms?.missing === 'false') {
        filterMissing = true
    }

    return {
        field,
        originalField: field,
        aggregation: {
            terms: {
                field: fieldKey,
                size: page * size,
            },
        },
        filterClause,
        filterExclude,
        filterMissing,
    }
}

const intervalFormat = (interval, param) => {
    switch (interval) {
        case 'year':
            return {
                gte: `${param}-01-01T00:00:00.000Z`,
                lte: `${param}-12-31T23:59:59.999Z`,
            }

        case 'month':
            return {
                gte: `${param}-01T00:00:00.000Z`,
                lte: `${param}-${daysInMonth(param)}T23:59:59.999Z`,
            }

        case 'week':
            return {
                gte: `${param}T00:00:00.000Z`,
                lt: `${DateTime.fromISO(param).plus({ days: 7 }).toISODate()}T23:59:59.999Z`,
            }

        case 'day':
            return {
                gte: `${param}T00:00:00.000Z`,
                lte: `${param}T23:59:59.999Z`,
            }

        case 'hour':
            return {
                gte: `${param}:00:00.000Z`,
                lte: `${param}:59:59.999Z`,
            }
    }
}

const buildHistogramField = (field, uuid, { interval = DEFAULT_INTERVAL, intervals } = {}, page = 1, size = DEFAULT_FACET_SIZE) => {
    const fieldKey = expandPrivate(field, uuid)

    let filterClause = null
    if (intervals?.include?.length) {
        filterClause = {
            bool: {
                should: intervals.include.map((selected) => ({
                    range: {
                        [fieldKey]: intervalFormat(interval, selected),
                    },
                })),
            },
        }
    }

    let filterMissing = null
    if (intervals?.missing === 'true') {
        filterMissing = false
    } else if (intervals?.missing === 'false') {
        filterMissing = true
    }

    return {
        field,
        originalField: field,
        aggregation: {
            date_histogram: {
                field: fieldKey,
                interval,
                min_doc_count: 1,
                order: { _key: 'desc' },
            },
            aggs: {
                bucket_truncate: {
                    bucket_sort: {
                        from: (page - 1) * size,
                        size,
                    },
                },
            },
        },
        filterClause,
        filterMissing,
    }
}

const rangeFormat = (fieldKey, range, isFilter = false) => {
    if (!range.includes('-')) {
        throw new Error('Invalid range format. Missing "-" sign.')
    }

    let rangeQuery
    const rangeEdges = range.split('-')

    if (!rangeEdges[0].length || !rangeEdges[1].length) {
        throw new Error('Invalid range format. Missing range edge.')
    }

    if (rangeEdges[0] === '*') {
        return {
            [isFilter ? 'lte' : 'to']: rangeEdges[1],
        }
    } else if (rangeEdges[1] === '*') {
        return {
            [isFilter ? 'gte' : 'from']: rangeEdges[0],
        }
    }

    return {
        [isFilter ? 'gte' : 'from']: rangeEdges[0],
        [isFilter ? 'lt' : 'to']: rangeEdges[1],
    }
}

const buildRangeField = (field, uuid, ranges) => {
    const fieldKey = expandPrivate(field, uuid)

    let filterClause = null
    if (ranges?.include?.length) {
        filterClause = {
            bool: {
                should: ranges.include.map((selected) => ({
                    range: {
                        [fieldKey]: rangeFormat(fieldKey, selected, true),
                    },
                })),
            },
        }
    }

    let filterExclude = null
    if (ranges?.exclude?.length) {
        filterExclude = {
            bool: {
                should: ranges.exclude.map((selected) => ({
                    range: {
                        [fieldKey]: rangeFormat(fieldKey, selected, true),
                    },
                })),
            },
        }
    }

    let filterMissing = null
    if (ranges?.missing === 'true') {
        filterMissing = false
    } else if (ranges?.missing === 'false') {
        filterMissing = true
    }

    return {
        field,
        originalField: field,
        aggregation: {
            range: {
                field: fieldKey,
                ranges: aggregationFields[field].buckets.map(({ key }) => ({
                    key,
                    ...rangeFormat(fieldKey, key),
                })),
            },
        },
        filterClause,
        filterExclude,
        filterMissing,
    }
}

const buildMissingField = (field, uuid) => ({
    field: `${field}-missing`,
    originalField: field,
    aggregation: {
        missing: { field: expandPrivate(field, uuid) },
    },
})

const prepareFilter = (field) => {
    const filter = []

    if (field.filterClause) {
        filter.push(field.filterClause)
    }

    if (field.filterExclude) {
        filter.push({ bool: { must_not: field.filterExclude } })
    }

    if (typeof field.filterMissing === 'boolean') {
        const exists = { exists: { field: field.field } }
        if (field.filterMissing) {
            filter.push(exists)
        } else {
            filter.push({ bool: { must_not: exists } })
        }
    }

    if (filter.length) {
        return filter
    }
}

const buildFilter = (fields) => {
    const filter = []

    fields.forEach((field) => {
        const fieldFilter = prepareFilter(field)

        if (fieldFilter) {
            if (aggregationFields[field.originalField].type === 'term-and') {
                filter.push(...fieldFilter)
            } else {
                filter.push({ bool: { should: fieldFilter } })
            }
        }
    })

    if (filter.length) {
        return { bool: { filter } }
    } else {
        return { bool: {} }
    }
}

const buildAggs = (fields) =>
    fields.reduce(
        (result, field) => ({
            ...result,
            [field.field]: {
                aggs: {
                    values: field.aggregation,
                    count: { cardinality: { field: field.field } },
                },
                filter: buildFilter(fields.filter((other) => other.originalField !== field.originalField)),
            },
        }),
        {}
    )

const getAggregationFields = (type, fieldList) =>
    Object.entries(aggregationFields)
        .filter(([, field]) => field.type.startsWith(type))
        .map(([key]) => key)
        .filter((field) => fieldList === '*' || (Array.isArray(fieldList) && fieldList.includes(field)))

const buildSearchQuery = (
    { q = '*', page = 1, size = 0, order, collections = [], facets = {}, filters = {} } = {},
    type = 'results',
    fieldList = '*',
    missing = false,
    searchFields,
    uuid
) => {
    const query = buildQuery(q, filters, searchFields)
    const sort = buildSortQuery(order)
    const significantFields = fieldList === '*' ? '*' : [...fieldList, ...Object.keys(filters)]

    const dateFields = getAggregationFields('date', significantFields)
    const termFields = getAggregationFields('term', significantFields)
    const rangeFields = getAggregationFields('range', significantFields)

    const fields = missing
        ? [
              ...dateFields.map((field) => buildMissingField(field, uuid)),
              ...termFields.map((field) => buildMissingField(field, uuid)),
              ...rangeFields.map((field) => buildMissingField(field, uuid)),
          ]
        : [
              ...dateFields.map((field) => buildHistogramField(field, uuid, filters[field], facets[field])),
              ...termFields.map((field) => buildTermsField(field, uuid, filters[field], facets[field])),
              ...rangeFields.map((field) => buildRangeField(field, uuid, filters[field], facets[field])),
          ]

    const postFilter = buildFilter(fields)
    const aggs = buildAggs(fields)

    const highlightFields = {}
    searchFields.highlight.forEach((field) => {
        highlightFields[field] = HIGHLIGHT_SETTINGS
    })

    return {
        from: (page - 1) * size,
        size: type === 'results' ? size : 0,
        query,
        sort,
        post_filter: postFilter,
        aggs: type === 'results' ? {} : aggs,
        collections,
        _source: searchFields._source,
        highlight: {
            fields: highlightFields,
        },
    }
}

export default buildSearchQuery
