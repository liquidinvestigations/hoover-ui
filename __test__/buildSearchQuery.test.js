import buildSearchQuery from '../src/backend/buildSearchQuery'

const searchFields = {
    all: [],
    _source: [],
    highlight: [],
}

it('builds a default query', () => {
    const query = buildSearchQuery({}, 'aggregations', '*', false, searchFields, [], 'testuser')
    expect(query).toMatchSnapshot()
})

it('builds a query with a filetype filter', () => {
    const query = buildSearchQuery(
        {
            filters: {
                filetype: {
                    include: ['email', 'pdf'],
                },
            },
        },
        'aggregations',
        '*',
        false,
        searchFields,
        [],
        'testuser',
    )

    expect(query.post_filter).toMatchObject({
        bool: {
            filter: [{ bool: { should: [{ terms: { filetype: ['email', 'pdf'] } }] } }],
        },
    })

    expect(query.aggs['email-domains']).toMatchObject({
        aggs: {
            values: {
                terms: { field: 'email-domains' },
            },
        },
        filter: {
            bool: {
                filter: [{ bool: { should: [{ terms: { filetype: ['email', 'pdf'] } }] } }],
            },
        },
    })

    expect(query.aggs.filetype).toMatchObject({
        aggs: {
            values: {
                terms: { field: 'filetype' },
            },
        },
    })
})

it('builds a query with a date histogram by years filter', () => {
    const query = buildSearchQuery(
        {
            filters: {
                date: {
                    intervals: {
                        include: ['2009'],
                    },
                },
            },
        },
        'aggregations',
        '*',
        false,
        searchFields,
        [],
        'testuser',
    )

    const yearFilter = {
        bool: {
            should: [
                {
                    bool: {
                        should: [
                            {
                                range: {
                                    date: {
                                        gte: '2009-01-01T00:00:00.000Z',
                                        lte: '2009-12-31T23:59:59.999Z',
                                    },
                                },
                            },
                        ],
                    },
                },
            ],
        },
    }

    expect(query.post_filter).toMatchObject({
        bool: {
            filter: [yearFilter],
        },
    })

    expect(query.aggs.filetype).toMatchObject({
        aggs: {
            values: {
                terms: { field: 'filetype' },
            },
        },
        filter: {
            bool: {
                filter: [yearFilter],
            },
        },
    })

    expect(query.aggs.date).toMatchObject({
        aggs: {
            values: {
                date_histogram: { field: 'date', interval: 'year' },
            },
        },
        filter: {},
    })
})

it('builds a query with multiple fields filtered', () => {
    const query = buildSearchQuery(
        {
            filters: {
                filetype: {
                    include: ['doc', 'email'],
                },
                'email-domains': {
                    include: ['gmail.com'],
                },
            },
        },
        'aggregations',
        '*',
        false,
        searchFields,
        [],
        'testuser',
    )

    expect(query.post_filter).toMatchObject({
        bool: {
            filter: [
                { bool: { should: [{ terms: { filetype: ['doc', 'email'] } }] } },
                { bool: { should: [{ terms: { 'email-domains': ['gmail.com'] } }] } },
            ],
        },
    })

    expect(query.aggs.filetype).toMatchObject({
        aggs: {
            values: {
                terms: { field: 'filetype' },
            },
        },
        filter: {
            bool: {
                filter: [{ bool: { should: [{ terms: { 'email-domains': ['gmail.com'] } }] } }],
            },
        },
    })

    expect(query.aggs['email-domains']).toMatchObject({
        aggs: {
            values: {
                terms: { field: 'email-domains' },
            },
        },
        filter: {
            bool: {
                filter: [{ bool: { should: [{ terms: { filetype: ['doc', 'email'] } }] } }],
            },
        },
    })
})
