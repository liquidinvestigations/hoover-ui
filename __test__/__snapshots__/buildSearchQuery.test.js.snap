// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`builds a default query 1`] = `
{
  "_source": [],
  "aggs": {
    "attachments": {
      "aggs": {
        "count": {
          "cardinality": {
            "field": "attachments",
          },
        },
        "values": {
          "terms": {
            "field": "attachments",
            "size": 44,
          },
        },
      },
      "filter": {
        "bool": {},
      },
    },
    "content-type": {
      "aggs": {
        "count": {
          "cardinality": {
            "field": "content-type",
          },
        },
        "values": {
          "terms": {
            "field": "content-type",
            "size": 44,
          },
        },
      },
      "filter": {
        "bool": {},
      },
    },
    "date": {
      "aggs": {
        "count": {
          "cardinality": {
            "field": "date",
          },
        },
        "values": {
          "aggs": {
            "bucket_truncate": {
              "bucket_sort": {
                "from": 0,
                "size": 44,
              },
            },
          },
          "date_histogram": {
            "field": "date",
            "interval": "year",
            "min_doc_count": 1,
            "order": {
              "_key": "desc",
            },
          },
        },
      },
      "filter": {
        "bool": {},
      },
    },
    "date-created": {
      "aggs": {
        "count": {
          "cardinality": {
            "field": "date-created",
          },
        },
        "values": {
          "aggs": {
            "bucket_truncate": {
              "bucket_sort": {
                "from": 0,
                "size": 44,
              },
            },
          },
          "date_histogram": {
            "field": "date-created",
            "interval": "year",
            "min_doc_count": 1,
            "order": {
              "_key": "desc",
            },
          },
        },
      },
      "filter": {
        "bool": {},
      },
    },
    "detected-objects.object.keyword": {
      "aggs": {
        "count": {
          "cardinality": {
            "field": "detected-objects.object.keyword",
          },
        },
        "values": {
          "terms": {
            "field": "detected-objects.object.keyword",
            "size": 44,
          },
        },
      },
      "filter": {
        "bool": {},
      },
    },
    "email-domains": {
      "aggs": {
        "count": {
          "cardinality": {
            "field": "email-domains",
          },
        },
        "values": {
          "terms": {
            "field": "email-domains",
            "size": 44,
          },
        },
      },
      "filter": {
        "bool": {},
      },
    },
    "entity-type.event.keyword": {
      "aggs": {
        "count": {
          "cardinality": {
            "field": "entity-type.event.keyword",
          },
        },
        "values": {
          "terms": {
            "field": "entity-type.event.keyword",
            "size": 44,
          },
        },
      },
      "filter": {
        "bool": {},
      },
    },
    "entity-type.location.keyword": {
      "aggs": {
        "count": {
          "cardinality": {
            "field": "entity-type.location.keyword",
          },
        },
        "values": {
          "terms": {
            "field": "entity-type.location.keyword",
            "size": 44,
          },
        },
      },
      "filter": {
        "bool": {},
      },
    },
    "entity-type.money.keyword": {
      "aggs": {
        "count": {
          "cardinality": {
            "field": "entity-type.money.keyword",
          },
        },
        "values": {
          "terms": {
            "field": "entity-type.money.keyword",
            "size": 44,
          },
        },
      },
      "filter": {
        "bool": {},
      },
    },
    "entity-type.organization.keyword": {
      "aggs": {
        "count": {
          "cardinality": {
            "field": "entity-type.organization.keyword",
          },
        },
        "values": {
          "terms": {
            "field": "entity-type.organization.keyword",
            "size": 44,
          },
        },
      },
      "filter": {
        "bool": {},
      },
    },
    "entity-type.person.keyword": {
      "aggs": {
        "count": {
          "cardinality": {
            "field": "entity-type.person.keyword",
          },
        },
        "values": {
          "terms": {
            "field": "entity-type.person.keyword",
            "size": 44,
          },
        },
      },
      "filter": {
        "bool": {},
      },
    },
    "entity.keyword": {
      "aggs": {
        "count": {
          "cardinality": {
            "field": "entity.keyword",
          },
        },
        "values": {
          "terms": {
            "field": "entity.keyword",
            "size": 44,
          },
        },
      },
      "filter": {
        "bool": {},
      },
    },
    "filetype": {
      "aggs": {
        "count": {
          "cardinality": {
            "field": "filetype",
          },
        },
        "values": {
          "terms": {
            "field": "filetype",
            "size": 44,
          },
        },
      },
      "filter": {
        "bool": {},
      },
    },
    "from.keyword": {
      "aggs": {
        "count": {
          "cardinality": {
            "field": "from.keyword",
          },
        },
        "values": {
          "terms": {
            "field": "from.keyword",
            "size": 44,
          },
        },
      },
      "filter": {
        "bool": {},
      },
    },
    "image-classes.class.keyword": {
      "aggs": {
        "count": {
          "cardinality": {
            "field": "image-classes.class.keyword",
          },
        },
        "values": {
          "terms": {
            "field": "image-classes.class.keyword",
            "size": 44,
          },
        },
      },
      "filter": {
        "bool": {},
      },
    },
    "is-table": {
      "aggs": {
        "count": {
          "cardinality": {
            "field": "is-table",
          },
        },
        "values": {
          "terms": {
            "field": "is-table",
            "size": 44,
          },
        },
      },
      "filter": {
        "bool": {},
      },
    },
    "lang": {
      "aggs": {
        "count": {
          "cardinality": {
            "field": "lang",
          },
        },
        "values": {
          "terms": {
            "field": "lang",
            "size": 44,
          },
        },
      },
      "filter": {
        "bool": {},
      },
    },
    "ocr": {
      "aggs": {
        "count": {
          "cardinality": {
            "field": "ocr",
          },
        },
        "values": {
          "terms": {
            "field": "ocr",
            "size": 44,
          },
        },
      },
      "filter": {
        "bool": {},
      },
    },
    "ocrimage": {
      "aggs": {
        "count": {
          "cardinality": {
            "field": "ocrimage",
          },
        },
        "values": {
          "terms": {
            "field": "ocrimage",
            "size": 44,
          },
        },
      },
      "filter": {
        "bool": {},
      },
    },
    "ocrpdf": {
      "aggs": {
        "count": {
          "cardinality": {
            "field": "ocrpdf",
          },
        },
        "values": {
          "terms": {
            "field": "ocrpdf",
            "size": 44,
          },
        },
      },
      "filter": {
        "bool": {},
      },
    },
    "path-parts": {
      "aggs": {
        "count": {
          "cardinality": {
            "field": "path-parts",
          },
        },
        "values": {
          "terms": {
            "field": "path-parts",
            "size": 44,
          },
        },
      },
      "filter": {
        "bool": {},
      },
    },
    "pgp": {
      "aggs": {
        "count": {
          "cardinality": {
            "field": "pgp",
          },
        },
        "values": {
          "terms": {
            "field": "pgp",
            "size": 44,
          },
        },
      },
      "filter": {
        "bool": {},
      },
    },
    "priv-tags": {
      "aggs": {
        "count": {
          "cardinality": {
            "field": "priv-tags",
          },
        },
        "values": {
          "terms": {
            "field": "priv-tags.testuser",
            "size": 44,
          },
        },
      },
      "filter": {
        "bool": {},
      },
    },
    "size": {
      "aggs": {
        "count": {
          "cardinality": {
            "field": "size",
          },
        },
        "values": {
          "range": {
            "field": "size",
            "ranges": [
              {
                "from": "0",
                "key": "0-1",
                "to": "1",
              },
              {
                "from": "1",
                "key": "1-1000",
                "to": "1000",
              },
              {
                "from": "1000",
                "key": "1000-500000",
                "to": "500000",
              },
              {
                "from": "500000",
                "key": "500000-1000000",
                "to": "1000000",
              },
              {
                "from": "1000000",
                "key": "1000000-500000000",
                "to": "500000000",
              },
              {
                "from": "500000000",
                "key": "500000000-1000000000",
                "to": "1000000000",
              },
              {
                "from": "1000000000",
                "key": "1000000000-*",
              },
            ],
          },
        },
      },
      "filter": {
        "bool": {},
      },
    },
    "skipped": {
      "aggs": {
        "count": {
          "cardinality": {
            "field": "skipped",
          },
        },
        "values": {
          "terms": {
            "field": "skipped",
            "size": 44,
          },
        },
      },
      "filter": {
        "bool": {},
      },
    },
    "table-col-count": {
      "aggs": {
        "count": {
          "cardinality": {
            "field": "table-col-count",
          },
        },
        "values": {
          "range": {
            "field": "table-col-count",
            "ranges": [
              {
                "from": "0",
                "key": "0-1",
                "to": "1",
              },
              {
                "from": "1",
                "key": "1-3",
                "to": "3",
              },
              {
                "from": "4",
                "key": "4-9",
                "to": "9",
              },
              {
                "from": "10",
                "key": "10-29",
                "to": "29",
              },
              {
                "from": "30",
                "key": "30-99",
                "to": "99",
              },
              {
                "from": "100",
                "key": "100-*",
              },
            ],
          },
        },
      },
      "filter": {
        "bool": {},
      },
    },
    "table-columns": {
      "aggs": {
        "count": {
          "cardinality": {
            "field": "table-columns",
          },
        },
        "values": {
          "terms": {
            "field": "table-columns",
            "size": 44,
          },
        },
      },
      "filter": {
        "bool": {},
      },
    },
    "table-row-count": {
      "aggs": {
        "count": {
          "cardinality": {
            "field": "table-row-count",
          },
        },
        "values": {
          "range": {
            "field": "table-row-count",
            "ranges": [
              {
                "from": "0",
                "key": "0-1",
                "to": "1",
              },
              {
                "from": "1",
                "key": "1-9",
                "to": "9",
              },
              {
                "from": "10",
                "key": "10-99",
                "to": "99",
              },
              {
                "from": "100",
                "key": "100-999",
                "to": "999",
              },
              {
                "from": "1000",
                "key": "1000-9999",
                "to": "9999",
              },
              {
                "from": "10000",
                "key": "10000-99999",
                "to": "99999",
              },
              {
                "from": "100000",
                "key": "100000-999999",
                "to": "999999",
              },
              {
                "from": "1000000",
                "key": "1000000-*",
              },
            ],
          },
        },
      },
      "filter": {
        "bool": {},
      },
    },
    "table-sheet-count": {
      "aggs": {
        "count": {
          "cardinality": {
            "field": "table-sheet-count",
          },
        },
        "values": {
          "range": {
            "field": "table-sheet-count",
            "ranges": [
              {
                "from": "0",
                "key": "0-1",
                "to": "1",
              },
              {
                "from": "1",
                "key": "1-3",
                "to": "3",
              },
              {
                "from": "4",
                "key": "4-9",
                "to": "9",
              },
              {
                "from": "10",
                "key": "10-29",
                "to": "29",
              },
              {
                "from": "30",
                "key": "30-99",
                "to": "99",
              },
              {
                "from": "100",
                "key": "100-*",
              },
            ],
          },
        },
      },
      "filter": {
        "bool": {},
      },
    },
    "table-sheets": {
      "aggs": {
        "count": {
          "cardinality": {
            "field": "table-sheets",
          },
        },
        "values": {
          "terms": {
            "field": "table-sheets",
            "size": 44,
          },
        },
      },
      "filter": {
        "bool": {},
      },
    },
    "tags": {
      "aggs": {
        "count": {
          "cardinality": {
            "field": "tags",
          },
        },
        "values": {
          "terms": {
            "field": "tags",
            "size": 44,
          },
        },
      },
      "filter": {
        "bool": {},
      },
    },
    "to.keyword": {
      "aggs": {
        "count": {
          "cardinality": {
            "field": "to.keyword",
          },
        },
        "values": {
          "terms": {
            "field": "to.keyword",
            "size": 44,
          },
        },
      },
      "filter": {
        "bool": {},
      },
    },
    "translated-from": {
      "aggs": {
        "count": {
          "cardinality": {
            "field": "translated-from",
          },
        },
        "values": {
          "terms": {
            "field": "translated-from",
            "size": 44,
          },
        },
      },
      "filter": {
        "bool": {},
      },
    },
    "translated-to": {
      "aggs": {
        "count": {
          "cardinality": {
            "field": "translated-to",
          },
        },
        "values": {
          "terms": {
            "field": "translated-to",
            "size": 44,
          },
        },
      },
      "filter": {
        "bool": {},
      },
    },
    "word-count": {
      "aggs": {
        "count": {
          "cardinality": {
            "field": "word-count",
          },
        },
        "values": {
          "range": {
            "field": "word-count",
            "ranges": [
              {
                "from": "0",
                "key": "0-1",
                "to": "1",
              },
              {
                "from": "1",
                "key": "1-9",
                "to": "9",
              },
              {
                "from": "10",
                "key": "10-99",
                "to": "99",
              },
              {
                "from": "100",
                "key": "100-999",
                "to": "999",
              },
              {
                "from": "1000",
                "key": "1000-9999",
                "to": "9999",
              },
              {
                "from": "10000",
                "key": "10000-99999",
                "to": "99999",
              },
              {
                "from": "100000",
                "key": "100000-999999",
                "to": "999999",
              },
              {
                "from": "1000000",
                "key": "1000000-*",
              },
            ],
          },
        },
      },
      "filter": {
        "bool": {},
      },
    },
  },
  "collections": [],
  "dedup_collections": undefined,
  "from": 0,
  "highlight": {
    "fields": {},
  },
  "post_filter": {
    "bool": {},
  },
  "query": {
    "query_string": {
      "default_operator": "AND",
      "fields": [],
      "lenient": true,
      "query": "*",
    },
  },
  "size": 0,
  "sort": [],
}
`;
